File: backend/app/services/uid_service.py
"""
uid_service.py — Cryptographic UID generation with collision retry.
Replaces: backend/src/services/UIDService.ts

Algorithm: secrets.token_bytes(6) → big-endian int → decimal string → first 12 chars → zero-pad
"""

import secrets
import re

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Share

UID_LENGTH = 6
UID_MODULO = 10**UID_LENGTH


def _generate_uid_candidate() -> str:
    """Generate a single 6-digit UID candidate from random bytes."""
    raw = secrets.token_bytes(4)
    num = int.from_bytes(raw, byteorder="big") % UID_MODULO
    return str(num).zfill(UID_LENGTH)


async def generate_uid(db: AsyncSession, max_retries: int = 10) -> str:
    """Generate a unique 6-digit UID, retrying on collision up to max_retries times."""
    for _ in range(max_retries):
        candidate = _generate_uid_candidate()
        existing = await db.execute(select(Share.id).where(Share.uid == candidate).limit(1))
        if existing.scalar_one_or_none() is None:
            return candidate
    raise RuntimeError("Failed to generate a unique UID after max retries")


def normalize_uid(raw: str) -> str:
    """Strip all non-digit characters."""
    return re.sub(r"\D", "", raw)[:UID_LENGTH]


def is_valid_uid(uid: str) -> bool:
    """Check if a string is exactly 6 digits."""
    return bool(re.fullmatch(rf"\d{{{UID_LENGTH}}}", uid))


def format_uid(uid: str) -> str:
    """Format a 6-digit UID into groups: '123 456'."""
    clean = re.sub(r"\D", "", uid)[:UID_LENGTH]
    parts = [clean[i : i + 3] for i in range(0, len(clean), 3)]
    return " ".join(parts)

File: backend/src/services/UIDService.ts
import crypto from 'crypto';
import { prisma } from '../db/prisma';

const MAX_ATTEMPTS = 10;

/**
 * Generate a cryptographically random 12-digit numeric UID.
 * Uses 6 random bytes (48 bits of entropy) converted to a decimal string.
 * Retries on collision (index-backed uniqueness check).
 */
export async function generateUID(): Promise<string> {
  for (let i = 0; i < MAX_ATTEMPTS; i++) {
    const bytes = crypto.randomBytes(6); // 48 bits of entropy
    const num = BigInt('0x' + bytes.toString('hex'));
    const uid = num.toString().slice(0, 12).padStart(12, '0');

    const existing = await prisma.share.findUnique({ where: { uid } });
    if (!existing) return uid;
  }
  throw new Error('UID generation failed after max retries — possible entropy exhaustion');
}

/**
 * Format a 12-digit UID into human-readable groups: "1234 5678 9012"
 */
export function formatUID(uid: string): string {
  return `${uid.slice(0, 4)} ${uid.slice(4, 8)} ${uid.slice(8, 12)}`;
}

/**
 * Strip formatting from a UID input: remove spaces, dashes, etc.
 */
export function normalizeUID(input: string): string {
  return input.replace(/[\s\-]/g, '');
}

/**
 * Validate that a string is a valid 12-digit numeric UID
 */
export function isValidUID(uid: string): boolean {
  return /^\d{12}$/.test(uid);
}
